lottie: fix & enhance parsing

mgrudzinska opened the pull request:
The object type key does not need to be provided
first. Previously, content was skipped until the
"ty" type key was found. If the key was placed at
the end, the entire object was ignored.
Now, before parsing subsequent elements, finding
the "ty" key is enforced. Since in-situ parsing
does not support backtracking, the key is searched 
with an internal function.

Introducing support for the "ty" key placed anywhere 
highlighted the use of redundant logic for parsing 
the list of shapes — this duplicate logic has been removed.

before:
<img width="400" alt="Zrzut ekranu 2024-05-30 o 01 42 45" src="https://github.com/thorvg/thorvg/assets/67589014/41c7292c-f9f5-4358-b3b2-1bf1403d409f">

after:
<img width="400" alt="Zrzut ekranu 2024-05-30 o 01 27 31" src="https://github.com/thorvg/thorvg/assets/67589014/6a1c1381-8fd1-41e9-85c7-c25de86111d5">

mgrudzinska commented:
file to test: [matte_bugs.lottie.json](https://github.com/thorvg/thorvg/files/15491840/matte_bugs.lottie.json) - from this issue: https://github.com/LottieFiles/dotlottie-web/issues/214
the other file may be private (?)

tinyjin commented on src/loaders/lottie/tvgLottieParser.cpp:
```diff
@@ -857,9 +861,9 @@ LottieRepeater* LottieParser::parseRepeater()
 {
-    auto type = getString();
+    if (!type) type = getString();
     if (!type) return nullptr;
```
In case `type` isn't nullptr from the beginning, we might be able to avoid unnecessary condition check by this code

```cpp
if (!type) {
    type = getString();
    if (!type) return nullptr;
}
```

tinyjin commented on src/loaders/lottie/tvgLottieParser.cpp:
```diff
@@ -887,6 +891,22 @@ LottieObject* LottieParser::parseObject()
     enterObject();
+
+    //object type key is not listed as the first one
+    if (peekValue() && strcmp(peekValue()->GetString(), "ty")) {
```
It could be more optimal and readable?
```cpp
auto value = peekValue()
if (value && strcmp(value->GetString(), "ty"))
```

hermet commented:
@mgrudzinska Could you please separate the `d` and `ty` patches? If the order of the `ty` patch is changed by other tools, it should be reviewed carefully.

mgrudzinska commented:
> @mgrudzinska Could you please separate the `d` and `ty` patches? If the order of the `ty` is changed by other tools/plugins (against the bodymovin), it should be reviewed carefully.

I separated them https://github.com/thorvg/thorvg/pull/2333
will rebase this one after 2333 is merged

hermet commented:
Seems this patch should be confirm & pushed first than https://github.com/thorvg/thorvg/pull/2333

mgrudzinska commented:
> Seems this patch should be confirm & pushed first than #2333

but what with the 'direction' ? In this pr I removed the whole section of parseShapes. I won't handle direction in group because it doesn;t make sense. I moved direction handling to separate shapes parsing, but you wanted this in a separate pr.

hermet commented:
Please cancel the draft after resolving the conflict. Thanks.

hermet commented on src/loaders/lottie/tvgLottieParser.cpp:
```diff
@@ -546,6 +546,7 @@ LottieRect* LottieParser::parseRect()
 LottieEllipse* LottieParser::parseEllipse()
 {
     auto ellipse = new LottieEllipse;
+    if (!ellipse) return nullptr;
```
@mgrudzinska just for your information, I intentionally removed the null check from the memory alloc fails.

mgrudzinska replied:
@hemet it makes perfect sense, since new will throw ex, not return null

mgrudzinska replied:
I added this, sorry, rebasing didn;t go well

hermet commented:
As you might know, the new usage could affect the performance of the Lottie format quite negatively. I think this is really bad structure for Lottie. Let's review this carefully.

mgrudzinska commented:
[skottie](https://skottie.skia.org/?h=128&w=128) rearranges the order of attributes and puts all "ty" tags at the end

hermet commented:
@mgrudzinska could you please rebase this?

hermet commented:
@mgrudzinska, could you please resubmit this PR from the ThorVG main repo?

mgrudzinska commented:
> @mgrudzinska, could you please resubmit this PR from the ThorVG main repo?

@hermet https://github.com/thorvg/thorvg/tree/mira/format rebased to main and on the main repo. 
if you prefer I can close this pr and open a new one, I don't see a way how to change a source branch of an already opened pr

mgrudzinska commented:
the same changes, just different branch: https://github.com/thorvg/thorvg/pull/3194
